### PR TITLE
FFD box fix for nonplanar faces

### PR DESCRIPTION
**Pull request by bigfooted**
FFD box fix to avoid ambiguous definition of nonplanar faces 

## Proposed Changes
*Give a brief overview of your contribution here in a few sentences.*
 The FFD box is defined using quadrilateral faces. These faces can in principle be non-planar. To avoid ambiguity, we define a supporting middle node and divide each face into 4 triangles. 

## Related Work
*Resolve any issues (bug fix or feature request), note any related PRs, or mention interactions with the work of others, if any.*

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with --warnlevel=3 when using meson).
- [x] My contribution is commented and consistent with SU2 style (https://su2code.github.io/docs_v7/Style-Guide/).
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp), if necessary.

### CODE REVIEW

**Comment by WallyMaier** [comment]
@bigfooted out of curiousity, how does a non-planar ffd box affect the deformation/optimization process?

**Comment by bigfooted** [comment]
> @bigfooted out of curiosity, how does a non-planar FFD box affect the deformation/optimization process?

When deforming a mesh, there can be mesh nodes close to the edge (planar sides) of the box that are either inside or outside of the box, depending on how you the nonplanar face ABCD is interpreted. This ambiguity can be resolved by either splitting the face into 2 triangles over a dividing line AC or BD, or by defining an average middle point and dividing the plane into 4 triangles.

File: Common/src/grid_movement/CFreeFormDefBox.cpp [hunk 1]
```diff
@@ -108,13 +108,17 @@ CFreeFormDefBox::~CFreeFormDefBox(void) {
         delete [] Coord_Control_Points[iOrder][jOrder][kOrder];
         delete [] ParCoord_Control_Points[iOrder][jOrder][kOrder];
         delete [] Coord_Control_Points_Copy[iOrder][jOrder][kOrder];
+        if (Coord_SupportCP != nullptr) delete [] Coord_SupportCP[iOrder][jOrder][kOrder];
```
Comment (pcarruscag) [inline Common/src/grid_movement/CFreeFormDefBox.cpp:111]: Is each of the dimensions allocated with 'new'? If so the middle arrays are being leaked.

File: Common/src/grid_movement/CFreeFormDefBox.cpp [hunk 2]
```diff
@@ -969,26 +973,38 @@ su2double *CFreeFormDefBox::GetParametricCoord_Iterative(unsigned long iPoint, s
...
+  // indices of the FFD box. Note that the front face is labelled 0,1,2,3 and the back face is 4,5,6,7
+  unsigned short Index[6][5] = {
+  {0,1,2,3,0},  // front side
+  {1,5,6,2,1},  // right side
+  {2,6,7,3,2},  // top side
+  {3,7,4,0,3},  // left side
+  {4,5,1,0,4},  // bottom side
+  {4,7,6,5,4}}; // back side
+
+/*--- The current approach is to subdivide each of the 6 faces of the hexahedral FFD box into 4 triangles
```
Comment (pcarruscag) [inline Common/src/grid_movement/CFreeFormDefBox.cpp:991]: Formatting

File: Common/src/grid_movement/CFreeFormDefBox.cpp [hunk 3]
```diff
@@ -1013,78 +1029,32 @@ bool CFreeFormDefBox::GetPointFFD(CGeometry *geometry, CConfig *config, unsigned
...
-  /*--- 1st tetrahedron {V0, V1, V2, V5}
-   2nd tetrahedron {V0, V2, V7, V5}
-   3th tetrahedron {V0, V2, V3, V7}
-   4th tetrahedron {V0, V5, V7, V4}
-   5th tetrahedron {V2, V7, V5, V6} ---*/
-
-  for (iVar = 0; iVar < 5; iVar++) {
-    Inside = true;
-    for (jVar = 0; jVar < 4; jVar++) {
+  // loop over the faces of the FFD box
```
Comment (pcarruscag) [inline Common/src/grid_movement/CFreeFormDefBox.cpp:1032]: /*---

**Review by pcarruscag — COMMENTED** [review]
Is checking the distance to planes sufficient? I guess the ffd is always a nice convex shape

File: Common/src/grid_movement/CFreeFormDefBox.cpp (hunk 1 above)
Comment (bigfooted) [reply Common/src/grid_movement/CFreeFormDefBox.cpp:111]: Ah, I should first `delete [] array[i[j][k]`, then `delete [] array[i][j]` then `delete [] array[i]` ?

Comment (pcarruscag) [reply Common/src/grid_movement/CFreeFormDefBox.cpp:111]: If all the dimensions are allocated with new, yes, unfortunately.

Comment (bigfooted) [reply Common/src/grid_movement/CFreeFormDefBox.cpp:111]: yep, looks like it, also for all the other variables there

File: Common/src/grid_movement/CFreeFormDefBox.cpp [hunk 4]
```diff
@@ -102,14 +102,25 @@ CFreeFormDefBox::CFreeFormDefBox(const unsigned short Degree[], unsigned short B
 CFreeFormDefBox::~CFreeFormDefBox(void) {
   unsigned short iOrder, jOrder, kOrder, iCornerPoints, iDim;
 
-  for (iOrder = 0; iOrder < lOrder; iOrder++)
-    for (jOrder = 0; jOrder < mOrder; jOrder++)
+  for (iOrder = 0; iOrder < lOrder; iOrder++){
```
Comment (pcarruscag) [inline Common/src/grid_movement/CFreeFormDefBox.cpp:105]:
```suggestion
  for (iOrder = 0; iOrder < lOrder; iOrder++) {
```

**Review by pcarruscag — APPROVED** [review]

**Pull request merged**